Implement "Use GitHub.com Style" option

thomasjo opened the pull request:
This PR aims to implement support for a *Use GitHub.com style* option that will ensure the previewed markdown document closely resembles what it would look like if rendered on github.com, instead of using the current UI theme's (markdown) style.

In order to keep the scope of the work in this PR reasonable, automating the process of keeping the GitHub Markdown styles updated, as well as styling fenced code blocks, will be handled in separate PRs.

- [x] Implement basic overrides for the lowest hanging fruit.
  - [x] Ensure that we correctly style *everything*, except fenced code blocks.
- [x] Add specs.
- [x] Move remaining work to separate tracking issue (automation task + fenced code block styles).
  https://github.com/atom/markdown-preview/issues/244

simurai commented:
> Is this even worth the effort?

I like it. Maybe even as the default? Not sure what the current default is based on.

thomasjo commented:
> Not sure what the current default is based on.

Right now its based on the color scheme of the active UI and syntax theme. We have discussed briefly about this being the default, and the consensus seemed to **not** make this the default since a white background might be visually jarring when using a dark theme (paraphrased).

simurai commented:
> Right now its based on the color scheme of the active UI and syntax theme.

Not for all themes. Just recently I added some dark styles to One dark that [overrides](https://github.com/atom/one-dark-ui/blob/master/styles/packages.less#L75-L137) the `markdown-preview` CSS. For example in Atom dark it still looks like this:

![screen shot 2015-04-12 at 3 54 48 pm](https://cloud.githubusercontent.com/assets/378023/7104668/491db2a2-e12c-11e4-9e12-0d3eab423c85.png)

So if a theme doesn't override the `markdown-preview` styles (make it dark), then it could look the same like on github.com.

simurai commented:
Actually.. the current markdown-preview already looks pretty close to github.com. Below I'm gonna paste the same as in the above screenshot. So just needs some small tweaks.. like link color, or the `<hr>` being solid.

# hello world
## hello world

you can write text [with links](http://example.com) inline or [link references][1].

* one _thing_ has *em* phasis
* two __things__ are **bold**

[1]: http://example.com

---

thomasjo commented:
A silly gif of how it looks with One Dark with and without the new option enabled

![github-style2](https://cloud.githubusercontent.com/assets/3622/7104793/8240c33c-e0f8-11e4-9afe-f4812135eb1a.gif)

> Not for all themes. Just recently I added some dark styles to One dark that overrides the markdown-preview CSS. For example in Atom dark[..]

And that's the "problem", UI themes can override how the markdown preview looks, and that's really what this PR is aimed at resolving by creating an option that provides a consistent look regardless of UI theme. That does mean we likely need to override everything since themes can override a lot of styles too. But I only started with the "low hanging fruits" since that provided some immediate benefits and were simple to target.

Fenced code blocks on the other hand, will be more difficult. It will basically require a full GitHub syntax theme, and that theme needs to be applied to the "private" `AtomEditor` instances used to render the code blocks.

I **really** would prefer to create a solution that mostly (if not entirely) automates the process of keeping the overrides in sync with github.com.

simurai commented:
I see. It's to enforce the GitHub styles, no matter if themes try to override it. Ok, yeah.

I wonder if there is a way to easily increase specificity without having two versions (the current markdown-preview styles AND the GitHub styles). Since they are so close and probably used to be the same at some point.

Maybe with a Less mixin:

```less
.markdown-preview {
  background: #fff;
  // rest of the markdown-preview styles
}

[data-use-github-style] {
  .markdown-preview();
}
```

If I'm not mistaken, this should output:

```css
.markdown-preview,
[data-use-github-style] .markdown-preview {
  background: #fff;
  // rest of the markdown-preview styles
}
```

So there would only be one version (preferably close to GitHub's) and it can be enforced with `[data-use-github-style]` like you already added a setting for it.

thomasjo commented:
I really like the mixin idea! Will test that out. That does mean that I need to ensure the default styling is a replica of github.com styling, which it is currently not (although close). I don't think that will be a problem though, seeing as only minimal changes are required. Thanks @simurai :bow:

thomasjo commented:
@simurai Your suggestion worked like a charm! I still need to figure out the syntax theme aspect, but worst case I can try to solve that aspect at a later stage.

simurai commented:
Nice :+1: 

Re. update the styles to match github.com. Would be cool to automate it, but not quite sure how. When inspecting, all CSS gets served in a single file. And AFAIK the source is written in Sass. But it seems that all relevant styles have a `.markdown-body` selector. Like `.markdown-body h1`. So to keep updating future-friendly we could do this:

1. Add the `.markdown-body` class to markdown-preview's markup.
2. Manually copy all the CSS that starts with `.markdown-body`. In DevTools you can also un-minify it to easier see where they start and stop. And then paste it into `atom/markdown-preview/styles/gfm.less` or so.. Then in a few months.. repeat!

But yeah.. could also be done in another PR to keep things separated.

jonrohan commented:
> update the styles to match github.com. Would be cool to automate it, but not quite sure how. When inspecting, all CSS gets served in a single file.

@thomasjo @simurai Oh boy are you in luck, we recently open sourced our `markdown-body` https://github.com/primer/markdown

benogle commented:
> Oh boy are you in luck, we recently open sourced our markdown-body https://github.com/primer/markdown

:boom:

simurai commented:
Sweet! It's :christmas_tree:

orderedlist commented:
What about using a shadow Dom for this, then inserting the primer markdown CSS? No need for overrides?

simurai commented:
The reason for the overriding is not to prevent leak in, but rather have it as an option in the settings. Some people prefer if the markdown-preview changes based on the theme (less jarring if it's a dark theme), but others prefer to always look like on github.com.

We could still use Shadow DOM to really make markdown-preview theming opt-in and not accidental. :+1:

thomasjo commented:
> Oh boy are you in luck, we recently open sourced our markdown-body https://github.com/primer/markdown

Awesome, awesome, awesome! :heart:

thomasjo commented:
> What about using a shadow Dom for this, then inserting the primer markdown CSS? No need for overrides?

As @simurai mentioned (which is not properly described in the PR body), this PR aims to add an option that allows users to ensure the GitHub styling is always applied, regardless of whether their current theme changes the appearance of the markdown preview or not.

I still like the idea of using the shadow DOM to prevent accidental spill over onto the markdown preview, so I will start looking into that.

thomasjo commented:
Updated the approach; now using the (almost) raw SCSS markdown component from Primer:
https://github.com/primer/markdown/blob/master/components/markdown.scss

I copied it our stylesheet almost verbatim. I only changed the `$margin` references to `@margin`. The copied styles exist under the `.markdown-body` selector just as in the original, and I then use this as a mixin within the `.markdown-preview` selector scope.

The only bits left in `.markdown-preview` now is stuff that we override to ensure proper output, in particular what is practically a duplicate of the `.highlight` stuff, converted so it can be applied to `atom-text-editor` elements. There are some selectors that only tweak margin and padding—I believe most of these can be :fire:, but I'll need to test some more to verify.

thomasjo commented:
I want to investigate dumping everything into shadow DOM to prevent "leakage"; anyone got any pointers? Never used it, so have no idea where to begin (other than heading to e.g. MDN or some such).

mnquintana commented:
@thomasjo Why not add primer/markdown as a dependency and use it more directly? (Other than the Sass / Less mismatch)

Although @jonrohan or someone else in the Primer org would need to publish it to npm.

thomasjo commented:
> Why not add primer/markdown as a dependency and use it more directly? (Other than the Sass / Less mismatch)

The Sass/Less mismatch is the problem here. I can't think of an easy way around that. 
If we really want to automate the process, then one idea is to grab the raw `.scss` file from GitHub, and pass it through some transformation to (naively) convert it to Less. Might be able to use e.g. [grunt-refactor](https://github.com/jonschlinkert/grunt-refactor) or just roll something dead simple with CoffeeScript.

thomasjo commented:
Oh, I plan on extracting the `.markdown-body` mixin into a separate file by the way. Just need to figure out a sensible location. Perhaps I'll reuse the `/assets` folder. If I place it in `/styles` it'll get automatically included in the DOM.

simurai commented:
There is also the already pre-preocessed CSS https://github.com/primer/markdown/blob/master/dist/user-content.css that could be imported. There is probably no need to change the variables.

thomasjo commented:
I'm reasonably sure the generated CSS cannot sensibly be used for mixin purposes. So if we go the route of using pre-processed CSS we lose some flexibility with regards to overriding styles at the Less level. I'm reasonably sure we'd need to have both `markdown-preview` and `markdown-body` classes applied to the preview container, and if we use that approach we put a lot of unnecessary complexity on theme authors if they desire to override styles IMHO; they'd potentially need to figure out whether which class selector is the applicable one.

Current approach is much simpler. If we want to automate the process, I believe grabbing the raw `.scss` file and transforming it into a `.less` file is the way to go.

simurai commented:
I would have to try to be sure, but Less has
- a [reference](http://lesscss.org/features/#import-options-reference) import option. Meaning, it wouldn't output the imported CSS, but you could use it inside a mixin.
- and a [less](http://lesscss.org/features/#import-options-less) option. It would make the imported CSS available to be used with Less. As a mixin for example.

So maybe this works:

```less
@import (reference, less) "user-content.css";

.markdown-preview {
  .markdown-body();
}
```

thomasjo commented:
It does not work as intended; it seems like the multiple `.markdown-preview` entries in `user-content.css` is not interpreted as a single mixin, and the result is that only the first entry is used. Once you start introducing descendant selectors and what not, it's no longer a mixin-able as far as I can tell.

In short, I only get this block mixed in

```css
.markdown-body {
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}
```

simurai commented:
Ohh.. you're right. Back to your original plan, I guess. Or just do it manually might be fine too. Probably won't change too frequently.

lee-dohm commented on spec/markdown-preview-spec.coffee:
```diff
@@ -348,3 +348,37 @@ describe "Markdown preview package", ->
+        expect(preview.element.getAttribute('data-use-github-style')).toBeNull()
+
+        atom.config.set 'markdown-preview.useGitHubStyle', true
+        expect(preview.element.getAttribute('data-use-github-style')).toBe ''
```
Should this just be `.not.toBeNull()` for consistency?

kevinsawicki commented on lib/markdown-preview-view.coffee:
```diff
@@ -124,6 +124,12 @@ class MarkdownPreviewView extends ScrollView
 
+    @disposables.add atom.config.observe 'markdown-preview.useGitHubStyle', (useGitHubStyle) =>
+      if useGitHubStyle
+        @element.setAttribute('data-use-github-style', '')
```
Super minor, but for data attributes, you could do `@element.dataset.useGithubStyle` instead of `setAttribute` with a `data-` prefix.

https://github.com/atom/atom/blob/1602b8cf45c8fb3c355691387fc4423750210f2c/src/pane-element.coffee#L84-L95
